clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/GoogleUserDataProvider.groovy

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.google.deploy.ops
+
+import com.netflix.frigga.Names
+import com.netflix.spinnaker.clouddriver.google.deploy.description.BasicGoogleDeployDescription
+import com.netflix.spinnaker.clouddriver.google.security.GoogleNamedAccountCredentials
+import groovy.transform.PackageScope
+import groovy.util.logging.Slf4j
+import org.springframework.stereotype.Component
+
+
+/**
+ * Provides the common user data from a local file to be applied to all Google deployments.
+ *
+ * Any custom user data specified for each deployment will be appended to common user data, allowing custom user data
+ * to override the common user data.
+ */
+@Slf4j
+@Component
+public class GoogleUserDataProvider {
+
+  GoogleUserDataProvider() {
+  }
+
+  /**
+   * Returns the user data as a Map.
+   */
+  Map getUserData(final String serverGroupName, final String instanceTemplateName,
+                  BasicGoogleDeployDescription description,
+                  GoogleNamedAccountCredentials credentials, String customUserData) {
+    String userDataFile = credentials.getUserDataFile()
+    String rawUserData = getFileContents(userDataFile)
+    String commonUserData = replaceTokens(rawUserData, description, serverGroupName, instanceTemplateName)
+
+    def userDataMap = stringToUserDataMap(commonUserData)
+
+    if (customUserData){
+      def customUserDataMap = stringToUserDataMap(customUserData)
+      userDataMap << customUserDataMap
+    }
+    userDataMap
+  }
+
+  /**
+   * Returns the contents of a file or an empty string if the file doesn't exist.
+   */
+  @PackageScope
+  String getFileContents(String filename) {
+    if (!filename) {
+      return ''
+    }
+    try {
+      File file = new File(filename)
+      def rawContentsList = file.readLines()
+      def contentsList = []
+      for (line in rawContentsList) {
+        if (!line.startsWith('#')) {
+          contentsList = contentsList << line
+        }
+      }
+      String contents = contentsList.join(',')
+      return contents
+    } catch (IOException e) {
+      log.warn("Failed to read user data file ${filename}; ${e.message}")
+      return ''
+    }
+  }
+  /**
+   * Returns the user data with the tokens replaced.
+   *
+   * Currently supports the following tokens:
+   *
+   * %%account%% 	    the name of the account
+   * %%accounttype%% 	the accountType of the account
+   * %%env%%        	the environment of the account
+   * %%app%%          the name of the app
+   * %%region%% 	    the deployment region
+   * %%group%% 	      the name of the server group
+   * %%autogrp%% 	    the name of the server group
```

Comment:
`autogrp` sounds like a reference to `ASG` in AWS - do we need to keep lit like this @duftler?